feat(participants-pane): Implement participants pane

gimre opened the pull request:
Implementation of a participants pane as a sidebar to the right of the meeting space. No new actions are available for now, only a consolidation of existing functionality.

<img width="997" alt="participants-pane" src="https://user-images.githubusercontent.com/4921311/112618871-48bec100-8e2f-11eb-9d8c-6e734c2b612b.png">

gimre commented:
Jenkins please test this please

gimre commented:
Jenkins please test this please

gimre commented:
Jenkins please test this please

saghul commented:
Jenkins please test this please.

saghul commented on react/features/participants-pane/constants.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Action type to signal the closing of the participants pane.
+ */
+export const PARTICIPANTS_PANE_CLOSE = 'PARTICIPANTS_PANE_CLOSE';
```
Please split these into actionTypes.js

saghul commented on react/features/participants-pane/functions.js:
```diff
@@ -0,0 +1,77 @@
+ * @param {any} value - Value.
+ * @returns {any} Initial value or null.
+ */
+export const renderable = value => value || null;
```
This is used in a single place, maybe you just need to apply Boolean() instead?

gimre replied:
Seems like I was trying to solve a problem that doesn't exist anymore... I had the impression that rendering anything other than `null` would cause the thing to get coerced into a string, resulting in "false" and "undefined" getting output into the DOM but it seems this not the case (anymore :s). Removed.

saghul commented on react/features/participants-pane/index.js:
```diff
@@ -0,0 +1,6 @@
+export * from './actions';
```
Drop this file altogether. Import specific things you need from each file / section. This avoids having weird import cycles that tend to break mobile.

saghul commented on react/features/participants-pane/reducer.js:
```diff
@@ -0,0 +1,36 @@
+ * Listen for actions that mutate the participants pane state
+ */
+ReducerRegistry.register(
+    REDUCER_KEY, (state = DEFAULT_STATE, action) => {
```
Unnecessary line break.

saghul commented on react/features/participants-pane/reducer.js:
```diff
@@ -0,0 +1,36 @@
+    REDUCER_KEY, (state = DEFAULT_STATE, action) => {
+        switch (action.type) {
+
+        case PARTICIPANTS_PANE_CLOSE:
```
Wouldn't it be better to have a single action which takes a parameter?

gimre replied:
I usually prefer being explicit but I can merge them if you like it better that way.

saghul commented on react/features/participants-pane/selectors.js:
```diff
@@ -0,0 +1,108 @@
+ * @param {Object} state - Global state.
+ * @returns {Object} Conference state.
+ */
+const getConferenceState = state => state['features/base/conference'];
```
These sectors should be exported in their respective features.

saghul commented on react/features/participants-pane/selectors.js:
```diff
@@ -0,0 +1,108 @@
+import { MEDIA_TYPE } from '../base/media';
```
I think we usually have selectors in functions.js. Can you please check if that is the usual case so we keep that consistency?

saghul commented on react/features/toolbox/components/web/Toolbox.js:
```diff
@@ -1140,6 +1152,25 @@ class Toolbox extends Component<Props> {
                     text = { t(`toolbar.${_raisedHand ? 'lowerYourHand' : 'raiseYourHand'}`) } />);
         }
 
+        if (this._shouldShowButton('participants-pane')) {
```
Check the former "invite" button too, for backwards compatibility.

saghul reviewed (changes requested):
That was something! Lots of goodness here Gabi, great work! Left a bunch of comments, PTAL.

@vp8x8 / @muscat1 I'd like some extra eyeballs on the CSS (including the styled objects) please.

gimre commented on css/filmstrip/_tile_view.scss:
```diff
@@ -16,7 +16,7 @@
         flex-direction: column;
         height: 100%;
-        width: 100vw;
+        width: 100%;
```
So, #filmstripRemoteVideos' container nicely takes up only available space, but then #filmstripRemoteVideos overflows out of it if there is anything taking up space next to it. Not sure if that make sense. 100vw was a happy coincidence here, 100% will accomodate content to the right.

vp8x8 commented on react/features/participants-pane/components/ParticipantsPane.js:
```diff
@@ -0,0 +1,61 @@
+import React, { useCallback } from 'react';
```
@flow

vp8x8 commented on react/features/participants-pane/components/RaisedHandIndicator.js:
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
```
@flow

vp8x8 commented on react/features/participants-pane/components/styled.js:
```diff
@@ -0,0 +1,331 @@
+    margin-top: 16px;
+  }
+
+  &::-webkit-scrollbar {
```
I don't think this works in Firefox. Is it OK to be used?

saghul replied:
Firefox should be supported the same way we support Chrome.

gimre replied:
There are exiting customizations done in scss for webkit scrollbars that mess with layout, am just disabling for Chrome. Firefox scrollbars are OK.

gimre commented on css/_videolayout_default.scss:
```diff
@@ -1,5 +1,14 @@
 #videoconference_page {
+    height: 100%;
     min-height: 100%;
```
I think the height addition was left over from another pass, will remove the change.

gimre commented on react/features/base/config/constants.js:
```diff
@@ -17,7 +17,7 @@ export const TOOLBAR_BUTTONS = [
     'fodeviceselection', 'hangup', 'profile', 'chat', 'recording',
     'livestreaming', 'etherpad', 'sharedvideo', 'settings', 'raisehand',
-    'videoquality', 'filmstrip', 'invite', 'feedback', 'stats', 'shortcuts',
+    'videoquality', 'filmstrip', 'participants-pane', 'feedback', 'stats', 'shortcuts',
```
Will do

gimre commented on react/features/base/icons/svg/mic-blocked.svg:
```diff
@@ -0,0 +1,3 @@
+<svg width="20" height="20" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
+<path fill-rule="evenodd" clip-rule="evenodd" d="M9.99992 1.66669C11.8409 1.66669 13.3333 3.15907 13.3333 5.00002V10C13.3333 11.5555 12.2678 12.8621 10.8269 13.23C10.8311 13.2638 10.8333 13.2983 10.8333 13.3334V14.9309C11.1452 14.8785 11.4474 14.7973 11.7369 14.69C12.0292 13.2018 13.2017 12.0293 14.6899 11.737C14.8904 11.196 14.9999 10.6108 14.9999 10C14.9999 9.53978 15.373 9.16669 15.8333 9.1666 ...
```
Missed this one

gimre commented on react/features/lobby/components/AbstractKnockingParticipantList.js:
```diff
@@ -4,6 +4,7 @@ import { PureComponent } from 'react';
 
 import { isLocalParticipantModerator } from '../../base/participants';
 import { setKnockingParticipantApproval } from '../actions';
+import { getLobby } from '../functions';
```
alrighty

gimre commented on react/features/participants-pane/components/MeetingParticipantItem.js:
```diff
@@ -0,0 +1,54 @@
+        <ParticipantItem
+            actionsTrigger = { ActionTrigger.Hover }
+            audioMuteState = { isAudioMuted ? MediaState.Muted : MediaState.Unmuted }
+            isRaised = { isRaised }
```
Changing to _isHighighted_ - was a bit iffy of that prop to begin with.

gimre commented on react/features/participants-pane/components/ParticipantItem.js:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * Media state for video
+     */
+    videoMuteState: MediaState,
```
Oh noes, @muscat1

gimre commented on react/features/participants-pane/components/VolumeSlider.js:
```diff
@@ -0,0 +1,69 @@
+  initialValue: number
+};
+
+export const VolumeSlider = ({
```
Oops, it is not, will remove it from the PR.

gimre commented on react/features/participants-pane/functions.js:
```diff
@@ -0,0 +1,77 @@
+ * @param {Array} arr2 - Second array.
+ * @returns {boolean} Are they equal or not.
+ */
+export const arrayEquality = (arr1, arr2) => {
```
Aye, removed.

saghul commented:
Some UI / UX review:

<img width="373" alt="Screenshot 2021-04-14 at 11 38 16" src="https://user-images.githubusercontent.com/317464/114689965-8cae3300-9d16-11eb-8f59-9a4f4237a0f0.png">

Most of those options don't apply to the local participant.

<img width="355" alt="Screenshot 2021-04-14 at 11 38 24" src="https://user-images.githubusercontent.com/317464/114690063-a64f7a80-9d16-11eb-81bc-75d8d3206048.png">

The Mute All button looks "disabled". Also there is no button to disable video for all, should there? The functionality exists.

<img width="339" alt="Screenshot 2021-04-14 at 11 39 20" src="https://user-images.githubusercontent.com/317464/114690233-cbdc8400-9d16-11eb-8d2c-37609845b1a8.png">

Video state is not displayed, just the mic status.

<img width="1927" alt="Screenshot 2021-04-14 at 11 40 39" src="https://user-images.githubusercontent.com/317464/114690336-de56bd80-9d16-11eb-9248-95a5790329eb.png">

Is it ok to have the lobby notifications in both places?

<img width="337" alt="Screenshot 2021-04-14 at 11 40 53" src="https://user-images.githubusercontent.com/317464/114690409-f595ab00-9d16-11eb-9588-eb29fc704e51.png">

The reject button has the same color as the background on hover, feels wrong.

<img width="1927" alt="Screenshot 2021-04-14 at 11 46 27" src="https://user-images.githubusercontent.com/317464/114690519-1100b600-9d17-11eb-8dfd-d13b48939d55.png">

The chat and participants panes have different backgrounds.

<img width="475" alt="Screenshot 2021-04-14 at 11 47 50" src="https://user-images.githubusercontent.com/317464/114690727-486f6280-9d17-11eb-8b9f-dd94b76cb7b6.png">

Not responsive. It should become a modal, like the chat pane.

saghul commented:
Looks like you forgot the `key` prop somewhere.

<img width="863" alt="Screenshot 2021-04-14 at 11 47 31" src="https://user-images.githubusercontent.com/317464/114690812-5a510580-9d17-11eb-81b4-a85c44a33e0c.png">

wernf commented:
Excellent job @gimre! I recently made some similar participants list. I integrated the participant pane in the chat pane for several reasons:
- I did not have to bother about any effect on the rendering of other elements, like large video and filmstrip
- in the case that both, the chat and the participants panes, are displayed the remaining size of the video pane is not too much reduced.
- I feel that participants list and chat fit well together as it is easy to address someone in the participants list to chat with.

Additionally I am hoping that in the future we could have the breakout rooms integrated in the same pane too. Something like this:

     ________________________________________________
    | [  Invite Someone (if alone in the meeting)  ] |
    | > Lobby (x)                                    |
    |  () Applicant             [ Admit ] [ Reject ] |
    |                                                |
    | > Main room (x)                       [ Join ] |
    |  () Participant        A-status V-status [...] |
    |  () Participant        A-status V-status [...] |
    |                                                |
    | [              Create Room                   ] |
    | > Room ABC (x)                        [ Join ] |
    |  () Participant        A-status V-status [...] |
    |  () Participant        A-status V-status [...] |
    |                                                |
    | ...                                            |
    |                                                |
    |                                                |
    |________________________________________________|
    |                                                |
    |  <Chat History>                                |
    |                                                |
    |  ____________________________________________  |
    | |                                            | |
    | | :) Type a message                       => | |
    | |____________________________________________| |
    |________________________________________________|

(Sorry for the ASCII art. I was just too lazy to make a proper mock-up ;) Each room is collapsible, only showing the room name and participants count for a more compact view. The participants can join the rooms themselves or can be moved by drag-and-drop between the various rooms.

So even without breakout rooms for now, it would make sense to have the participants and chat together in one pane.

saghul commented:
> So even without breakout rooms for now, it would make sense to have the participants and chat together in one pane.

We have plans to unify them. One thing at a time though :-) Thanks for the feedback!

PS: Love the ascii diagram!

gimre commented:
@saghul addressed all the comments - thanks for taking the time to thoroughly review this, I know it was quite a bit code. I also added a keyboard shortcut for the pane (P) :)

gimre commented:
Jenkins please test this please.

saghul reviewed (approved):
🚀
